compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -515,6 +516,23 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
     /// Type check a `let` statement.
     pub fn check_decl_local(&self, local: &'tcx hir::Local<'tcx>) {
+        // Check for destructuring assignment.
+        match local.source {
+            hir::LocalSource::AssignDesugar(eq_sign_span)
+                if !self.tcx.features().destructuring_assignment =>
+            {
+                feature_err(
+                    &self.tcx.sess.parse_sess,
+                    sym::destructuring_assignment,
+                    eq_sign_span,
+                    "destructuring assignments are unstable",
+                )
+                .span_label(local.pat.span, "cannot assign to this expression")
+                .emit();
+            }
+            _ => {}
+        }
```

Comment:
This seems like a pretty random place to report the feature gate error.
You could report it in AST lowering together with performing the actual desugaring instead.

Reply:
I agree it's random. In the lowering context, I would have to use this function to acess the features, I think: https://github.com/rust-lang/rust/blob/b2025326088b54fb3f083bebeba14e0a15bf00d3/compiler/rustc_session/src/session.rs#L642
However, it warns that one should use a `TyCtxt` whenever possible, which isn't available in the lowering context. Hence I put it in the above file. I don't understand this very well, so if you say it's fine to use the function `features_untracked` despite the warning, I'll do that.

Reply:
If `TyCtxt` is available then yes, `features_untracked` should not be used, but if it's not available (because it's early compilation), then using it is totally ok.
See other examples of feature gating with `features_untracked` in `rustc_ast_lowering`.

Reply:
Thanks for clarifying that! I've moved the feature error to lowering.